### PR TITLE
[dev-env] Introuces update to change existing environment

### PR DESCRIPTION
**Pull request by pschoffer**
## Description

This command is usefull when want to tweak existing environment. Common scenarios are:
* Enable/disable phpmyadmin/statsd/xdebug
* Switch wp version

## Steps to Test

1) Create environment
2) Run `vip dev-env update`
3) Make sure it does what you would expect it to do and correct default values are fetched from existing env

### CODE REVIEW

File: src/lib/dev-environment/dev-environment-core.js [hunk 1]
```diff
@@ -308,7 +341,7 @@ export async function getApplicationInformation( appId: number, envType: string
 	return appData;
 }
 
-export async function resolveImportPath( slug: string, fileName: string, searchReplace: string, inPlace: boolean ): Promise<SQLImportPaths> {
+export async function resolveImportPath( slug: string, fileName: string, searchReplace: String | String[], inPlace: boolean ): Promise<SQLImportPaths> {
```
Comment (pauarge) [inline src/lib/dev-environment/dev-environment-core.js]: `string` should be lowercase

Comment (pschoffer) [reply src/lib/dev-environment/dev-environment-core.js]: lowercase causes flow issue here: https://github.com/Automattic/vip/blob/fbb6502fea2304bdd69dfc97623cb41fd2d064be/src/lib/search-and-replace.js#L128

But maybe we should change type there instead of here.

Comment (pschoffer) [reply src/lib/dev-environment/dev-environment-core.js]: Fixed it at the source ...

File: src/bin/vip-import-sql.js [hunk 2]
```diff
@@ -506,6 +506,7 @@ Processing the SQL import for your environment...
 				environmentId: env.id,
 				basename: basename,
 				md5: md5,
+				searchReplace: [],
```
Comment (pauarge) [inline src/bin/vip-import-sql.js:509]: Are this changes related?

File: src/lib/dev-environment/dev-environment-cli.js [hunk 3]
```diff
@@ -270,52 +265,76 @@ export async function promptForComponent( component: string, allowLocal: boolean
...
-			choices,
+			choices: modChoices,
 			initial: initialModeIndex,
 		} );
 
 		modeResult = await select.run();
 	}
 
 	const messagePrefix = selectMode ? '\t' : `${ componentDisplayName } - `;
+	// path
```
Comment (pauarge) [inline src/lib/dev-environment/dev-environment-cli.js]: Leftover?

File: src/lib/dev-environment/types.js [hunk 4]
```diff
@@ -0,0 +1,54 @@
+
+export interface InstanceOptions {
+	title?: string,
+	multisite?: boolean,
```
Comment (pauarge) [inline src/lib/dev-environment/types.js]: Aren't title and multisite always defined?

File: src/lib/dev-environment/types.js [hunk 5]
```diff
@@ -0,0 +1,54 @@
+
```
Comment (pauarge) [inline src/lib/dev-environment/types.js:1]: General comment on this type definitions, which cascade all over the PR -- could we try to have less nullable types? It would save us a lot of `x?.something || ' '`.

File: src/bin/vip-import-sql.js (hunk 2 above)
Comment (pschoffer) [reply src/bin/vip-import-sql.js:509]: Not really, they just break flow validations.

File: src/lib/dev-environment/dev-environment-cli.js (hunk 3 above)
Comment (pschoffer) [reply src/lib/dev-environment/dev-environment-cli.js]:
```suggestion
```

File: src/lib/dev-environment/dev-environment-core.js [hunk 6]
```diff
@@ -108,12 +100,41 @@ export async function createEnvironment( instanceData: NewInstanceData ) {
...
+		throw new Error( 'Environment doesn\'t exist.' );
+	}
+
+	const preProcessedInstanceData = preProcessInstanceData( instanceData );
+
+	await prepareLandoEnv( preProcessedInstanceData, instancePath );
+}
+
+function preProcessInstanceData( instanceData: InstanceData ): InstanceData {
+	const newInstanceData = {
```
Comment (pauarge) [inline src/lib/dev-environment/dev-environment-core.js:128]: Does it have any side effects to modify the original object instead of having to create a new one?

Comment (pschoffer) [reply src/lib/dev-environment/dev-environment-core.js:128]: If you do:
```
const a = {};
const b = preprocess(a)
```
You don't expect a to be changed as well. [Pure functions ](https://en.wikipedia.org/wiki/Pure_function) are kinda easy to think about I feel like.

File: src/bin/vip-dev-env-update.js [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+
+		const defaultOptions: InstanceOptions = {
+			clientCode: currentInstanceData.clientCode.dir || currentInstanceData.clientCode.tag || 'latest',
+			muPlugins: currentInstanceData.muPlugins.dir || currentInstanceData.muPlugins.tag || 'latest',
+			wordpress: currentInstanceData.wordpress.tag,
+			elasticsearch: currentInstanceData.elasticsearch,
+			mariadb: currentInstanceData.mariadb,
+			statsd: currentInstanceData.statsd,
+			phpmyadmin: currentInstanceData.phpmyadmin,
+			xdebug: currentInstanceData.phpmyadmin,
```
Comment (pauarge) [inline src/bin/vip-dev-env-update.js]: Looks like this should be `.xdebug`

File: src/lib/dev-environment/dev-environment-core.js (hunk 6 above)
Comment (pauarge) [reply src/lib/dev-environment/dev-environment-core.js:128]: Agreed, but copying an object over and over has an effect on performance. My question comes from the fact that if what we're doing here is taking an object and putting it through a series of functions that modify it (and those functions are only used here), it's fine to change the object that is passed by reference.

This is a more philosophical comment, so not worth blocking the PR because of it.

File: src/bin/vip-dev-env-update.js [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+		debug( 'Read instance data', currentInstanceData );
+
+		const preselectedOptions = {
+			// Title and multisite can't be changed during update
+			title: currentInstanceData.wpTitle,
+			multisite: currentInstanceData.multisite,
+			...opt,
+		};
+
+		const defaultOptions: InstanceOptions = {
```
Comment (pauarge) [inline src/bin/vip-dev-env-update.js:58]: Maybe we could extract the default options (plus some validation that's happening here) outside the command itself.

**Review by pauarge — APPROVED** [review]
